x-pack/plugins/ml/public/settings/calendar/edit/import_modal.js

```diff
@@ -0,0 +1,289 @@
+            description: description[0].value,
+            start_time: start[0].value.valueOf(),
+            end_time: end[0].value.valueOf(),
+            asterisk: recurring
+          });
+        }
+      }
+    });
+    return mlEvents;
+  }
+
+  static parseICSFile = (data) => {
+    const cal = icalendar.parse_calendar(data);
+    return ImportModal.createEvents(cal);
+  }
+
+  static filterEvents = (events) => {
+    const now = moment().valueOf();
+    return events.filter(e => e.start_time > now);
+  }
+
+  // move to utils?
+  static readFile = (file) => {
+    return new Promise((resolve, reject) => {
+      if (file && file.size) {
+        const reader = new FileReader();
+        reader.readAsText(file);
+
+        reader.onload = (() => {
+          return () => {
+            const data = reader.result;
+            if (data === '') {
+              reject();
+            } else {
+              resolve({ data });
+            }
+          };
+        })(file);
+      } else {
+        reject();
+      }
+    });
+  }
+
+  handleImport = async (loadedFile) => {
+    const incomingFile = loadedFile[0];
+    const errorMessage = 'Could not parse ICS file.';
+    let events = [];
+
+    if (incomingFile && incomingFile.size <= (MAX_FILE_SIZE_MB * 1000000)) {
+      this.setState({ fileLoading: true, fileLoaded: true });
+
+      try {
+        const parsedFile = await ImportModal.readFile(incomingFile);
+        events = ImportModal.parseICSFile(parsedFile.data);
+
+        this.setState({
+          allImportedEvents: events,
+          fileLoading: false,
+          errorMessage: null,
+          includePastEvents: false
+        });
+      } catch (error) {
+        console.log(errorMessage, error);
+        this.setState({ errorMessage, fileLoading: false });
+      }
+    } else if (incomingFile && incomingFile.size > (MAX_FILE_SIZE_MB * 1000000)) {
+      this.setState({ fileLoading: false, errorMessage });
+    } else {
+      this.setState({ fileLoading: false, errorMessage: null });
+    }
+  }
+
+  onEventDelete = (eventId) => {
+    this.setState(prevState => ({
+      allImportedEvents: prevState.allImportedEvents.filter(event => event.event_id !== eventId),
+    }));
+  }
+
+  onCheckboxToggle = (e) => {
+    this.setState({
+      includePastEvents: e.target.checked,
+    });
+  };
+
+  handleEventsAdd = () => {
+    const events = this.selectedEvents.map((event) => ({
+      description: event.description,
+      start_time: event.start_time,
+      end_time: event.end_time
+    }));
+
+    this.props.addImportedEvents(events);
+  }
+
+  renderCallout = () => (
+    <EuiCallOut color="danger">
+      <p>{this.state.errorMessage}</p>
+    </EuiCallOut>
+  );
+
+  renderImportedEvents = () => {
+    const {
+      allImportedEvents,
+      includePastEvents
+    } = this.state;
+
+    let showRecurringWarning = false;
+
+    if (includePastEvents) {
+      this.selectedEvents = allImportedEvents;
```

Comment:
the setting of `selectedEvents` in this function looks like a side effect to me.
this `if` statement and the creation of `selectedEvents` could be moved to `render`.
`selectedEvents` could then be a local var and be passed into `renderCallout ` and `handleEventsAdd`.